cartridge

```diff
@@ -1,13 +1,88 @@
-package.path = package.path .. ';' .. app_dir .. '/.rocks/share/tarantool/?.lua'
-package.path = package.path .. ';' .. app_dir .. '/.rocks/share/tarantool/?/init.lua'
 
-local cli = require('cartridge-cli')
+local script_dir = fio.abspath(fio.dirname(arg[0]))
+local tarantool_dir = fio.abspath(fio.dirname(arg[-1]))
+
+local CLI_MODNAME = 'cartridge-cli'
+
+
+-- There is a little (or not) hack:
+--
+-- On require('cartridge-cli') we should load modules in that priority:
+-- - (1) in the same dir with this script (cartridge-cli development)
+-- - (2) rocks from Tarantool Enterprise SDK (using cartridge from SDK)
+-- - (3) like a usual module (from .rocks etc.)
+--
+-- This function is a custom loader for `cartridge-cli` module.
+-- It's inserted in `package.loaders` table on position 1 and affects
+-- only `cartridge-cli` module.
+--
+-- When `require(<modname>)` is called, Lua searches modules this way
+-- (https://www.lua.org/manual/5.1/manual.html#pdf-package.loaders):
+-- - in a package.preload table;
+-- - looks across paths from package.path;
+-- - looks across paths from package.cpath;
+-- - ans so on...
+--
+-- But Tarantool changes this behaviour by inserting custom loaders in this table
+-- (https://github.com/tarantool/tarantool/blob/4e27b0e61747c5c6a8e7b9e53b18fa78f24ae61a/src/lua/init.lua#L215)
+-- In fact, there is some other loader before `package.preload`, so the
+-- package.path changes don't affect searching module.
+--
+-- This loader makes `cartridge-cli` module loading priority independent
+-- from the Tarantool `package.loaders` and more clear for developers
+--
+local function cli_loader(modname)
+    -- affects only `cartridge-cli`
+    if modname ~= CLI_MODNAME then
+        return string.format('This module is not `%s', CLI_MODNAME)
+    end
+
+    -- possible module filenames
+    local possible_filenames = {
+        string.format('%s.lua', CLI_MODNAME),
+        string.format('%s/init.lua', CLI_MODNAME),
+    }
+
+    -- the directories to searche in a priority order
+    local possible_basepaths = {
+        -- (1) cartridge-cli development
+        script_dir,
+        -- (2) Tarantool Enterprise
+        fio.pathjoin(tarantool_dir, '.rocks/share/tarantool'),
+    }
+
+    local cli_filepath
+    local errors = {}
+
+    -- find the first existing file
+    for _, basepath in ipairs(possible_basepaths) do
+        for _, filename in ipairs(possible_filenames) do
+            local filepath = fio.pathjoin(basepath, filename)
+            if fio.path.exists(filepath) then
+                cli_filepath = filepath
+                break
+            else
+                table.insert(errors, string.format('no file %s', filepath))
+            end
+        end
+        if cli_filepath ~= nil then break end
+    end
+
+    if cli_filepath == nil then
+        return table.concat(errors, '\n')
+    end
+
+    return function()
+        return dofile(cli_filepath)
```

Comment:
Please consider to check several things:

- This file has correct lua sytnax (load(filename) should return correct result)
- This file is really cartridge-cli application

Reply:
I think that this loader behaviour should be the same as standard loaders have, so I will study how do standard loaders handle this situation and do the same. Are you OK with it?

Maybe it's enough to simply load `cartridge-cli` module?

Reply:
> Maybe it's enough to simply load cartridge-cli module?

Could you provide a correct and understandable for users message and test such behaviour?
But personally I don't want to see as you drop all predefined loaders.

> I will study how do standard loaders handle this situation and do the same

Agree. I like this way.